### PR TITLE
Add AwsAttributePropagatingSpanProcessor component to AWS X-Ray

### PR DESCRIPTION
**Pull request by bjrara**
## Context

`AwsAttributePropagatingSpanProcessor` is designed to handle attributes `aws.remote.application`, `aws.remote.operation` and `aws.local.operation` which will later be used to generate metrics under the background of https://github.com/open-telemetry/opentelemetry-java-contrib/issues/789.

It has a prerequisite PR https://github.com/open-telemetry/opentelemetry-java-contrib/pull/800.

## Description

`AwsAttributePropagatingSpanProcessor` accomplishes two things:

### 1. Propagating remote attributes.
The `aws.remote.application` and `aws.remote.operation` attributes are designed to be set by users into local client spans through manual instrumentation. However, since client spans are mostly generated by auto-instrumentation libraries which make them unaccessible to application developers.

[OTEP-207](https://github.com/open-telemetry/oteps/pull/207) seems to be a general solution that could help address this requirement, but it hasn't been finalized yet. We took the idea from [this comment](https://github.com/open-telemetry/oteps/pull/207#pullrequestreview-1006473604) which suggests to let SpanProcessor to do the propagation work. 

`AwsAttributePropagatingSpanProcessor ` is used as an alternative solution that inherit `aws.remote.application` and `aws.remote.operation` attributes from parent span to the child spans if present.

### 2. Setting and propagating local attributes.
`SpanMetricsProcessor` from https://github.com/open-telemetry/opentelemetry-java-contrib/issues/789 requires `aws.local.operation` (the initial entry point of a execution) to be injected into the local client spans to generate metrics that can later be used for other correlations.

`AwsAttributePropagatingSpanProcessor` sets this attribute with the span name of the local root span if the request is from remote, and passes it to child spans.

## Testing

JUnit test is created for this new component in the package.

## Documentation

It's an independent component that would require manual installation in the TracerProvider. We'll create a doc to describe the usage.

## Outstanding items

NONE.

### CODE REVIEW

**Review by willarmiros — APPROVED** [review]
This change LGTM and I support it being included in the X-Ray contrib repo.

@open-telemetry/java-instrumentation-approvers PTAL and merge when ready :)

File: aws-xray/src/main/java/io/opentelemetry/contrib/awsxray/RemoteAttributesSpanProcessor.java
Comment (jack-berg) [inline aws-xray/src/main/java/io/opentelemetry/contrib/awsxray/RemoteAttributesSpanProcessor.java:20]: Reminder that this artifact is stable and we can't make breaking changes to any of the API surface area. Therefore, we should think carefully about this addition.

Comments:
- Make the class final
- Add a private constructor
- Add a `public static RemoteAttributesSpanProcessor create()` method for initializing.

Comment (bjrara) [reply aws-xray/src/main/java/io/opentelemetry/contrib/awsxray/RemoteAttributesSpanProcessor.java]: Updated.

File: aws-xray/src/main/java/io/opentelemetry/contrib/awsxray/RemoteAttributesSpanProcessor.java
Comment (jack-berg) [inline aws-xray/src/main/java/io/opentelemetry/contrib/awsxray/RemoteAttributesSpanProcessor.java:42]: Can you add test coverage for the case where `aws.remote.application` / `aws.remote.operation` are null?

**Review by jack-berg — APPROVED** [review]
This looks ok to me. Note to other reviews that this artifact is stable so please take a close look at the new API surface area.

File: aws-xray/src/main/java/io/opentelemetry/contrib/awsxray/RemoteAttributesSpanProcessor.java
Comment (bjrara) [reply aws-xray/src/main/java/io/opentelemetry/contrib/awsxray/RemoteAttributesSpanProcessor.java:42]: Thanks for the review. Added a new test case for null check.

**Review by willarmiros — APPROVED** [review]
Additional changes LGTM! This can be merged whenever and thanks for the review @jack-berg

**Comment by jack-berg** [comment]
@open-telemetry/java-contrib-maintainers PTAL.

**Comment by bjrara** [comment]
Put this PR on hold as the component might need to extend its scope to cooperate with https://github.com/open-telemetry/opentelemetry-java-contrib/issues/789.

**Comment by thpierce** [comment]
Because the author of this PR is not currently available, I have created a new PR here: https://github.com/open-telemetry/opentelemetry-java-contrib/pull/856

Please close this PR if you are able.

**Pull request closed without merging**
